src/socket/udp_socket.cr

```diff
@@ -88,4 +88,139 @@ class UDPSocket < IPSocket
+    when Family::INET
+      optname = LibC::IP_MULTICAST_LOOP
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_LOOP
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    getsockopt_bool optname, level
+  end
+
+  # Sets whether transmitted multicast packets should be copied and sent back
+  # to the originator, if the host has joined the multicast group.
+  def multicast_loopback=(val : Bool)
+    case @family
+    when Family::INET
+      optname = LibC::IP_MULTICAST_LOOP
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_LOOP
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    setsockopt_bool optname, val, level
+  end
+
+  # Returns the current value of the `hoplimit` field on uni-cast packets.
+  # Datagrams with a `hoplimit` of `1` are not forwarded beyond the local network.
+  # Multicast datagrams with a `hoplimit` of `0` will not be transmitted on any
+  # network, but may be delivered locally if the sending host belongs to the
+  # destination group and multicast loopback is enabled.
+  def multicast_hops
+    case @family
+    when Family::INET
+      optname = LibC::IP_MULTICAST_TTL
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    getsockopt optname, 0, level
+  end
+
+  # The multicast hops option controls the `hoplimit` field on uni-cast packets.
+  # If `-1` is specified, the kernel will use a default value.
+  # If a value of `0` to `255` is specified, the packet will have the specified
+  # value as `hoplimit`. Other values are considered invalid.
+  # Datagrams with a `hoplimit` of `1` are not forwarded beyond the local network.
+  # Multicast datagrams with a `hoplimit` of `0` will not be transmitted on any
+  # network, but may be delivered locally if the sending host belongs to the
+  # destination group and multicast loopback is enabled.
+  def multicast_hops=(val : Int)
+    case @family
+    when Family::INET
+      optname = LibC::IP_MULTICAST_TTL
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    setsockopt optname, val, level
+    val
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv4 interface address for subsequent transmissions. Setting the interface
+  # to `0.0.0.0` will select the default interface.
+  def multicast_interface(address : IPAddress)
+    if @family == Family::INET
+      setsockopt LibC::IP_MULTICAST_IF, address.addr4.not_nil!, LibC::IPPROTO_IP
```

Comment:
When `address` is a not a v4 address, this would raise `NilAssertionFailed`. Could you make sure that `address.addr4` is valid and otherwise raise a custom error?